File: leetcode/3202.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        nums = [num%k for num in nums]
        maxx = 0

        for i in range(k):
            l = [0] * k
            ok = [True] * k #ok means want alternation (not i)
            for num in nums:
                if num == i:
                    for j in range(k):
                        if ok[j]:
                            l[j] += 1
                            if i != j:
                                ok[j] = False
                else:
                    if not ok[num]:
                        l[num] += 1
                        ok[num] = True
            maxx = max(maxx, max(l))
        
        return maxx

File: leetcode/3202.cpp
class Solution
{
public:
    int maximumLength(vector<int> &nums, int k)
    {
        int n = nums.size();
        int ret = 0;
        vector<vector<int>> dp(k, vector<int>(k, 0));
        for (int i = 0; i < n; ++i)
        {
            int current_remainder = nums[i] % k;
            for (int previous_remainder = 0; previous_remainder < k; ++previous_remainder)
            {
                dp[current_remainder][previous_remainder] =
                    max(dp[current_remainder][previous_remainder],
                        dp[previous_remainder][current_remainder] + 1);
                ret = max(ret, dp[current_remainder][previous_remainder]);
            }
        }
        return ret;
    }
};
